### PR TITLE
Remove deprecated "Errors" types on next-branch

### PR DESCRIPTION
**Pull request by jikim-msft**
#### Description

[5467](https://dev.azure.com/fluidframework/internal/_workitems/edit/5467/)

Deprecate error types on the next branch.

### CODE REVIEW

File: packages/common/driver-definitions/src/driverError.ts
Comment (Josmithr) [inline packages/common/driver-definitions/src/driverError.ts:111]: While we're here.
```suggestion
 * @remarks
 * Not expected to be implemented by a class or an object literal, but rather used in place of
```

File: packages/common/driver-definitions/src/driverError.ts
Comment (Josmithr) [inline packages/common/driver-definitions/src/driverError.ts:115]:
```suggestion
 * It will be either {@link DriverErrorTypes} or the specific driver's specialized error type enum,
```

File: packages/drivers/odsp-driver-definitions/src/factory.ts
Comment (Josmithr) [inline packages/drivers/odsp-driver-definitions/src/factory.ts:12]:
```suggestion
	 * @remarks
	 * If specified, SPO will fail snapshot request with 413 error (see {@link OdspErrorTypes.snapshotTooBig})
```

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (Josmithr) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:430]: I believe the uses of `DriverErrorTypes` here can be replaced with `OdspErrorTypes`, as the later should include all the flags of the former. This should be applicable to usages across this package, and similarly applicable for `routerlicious-driver`.

File: packages/loader/driver-utils/src/network.ts
Comment (Josmithr) [inline packages/loader/driver-utils/src/network.ts:46]:
```suggestion
	/**
	 * {@inheritDoc IFluidErrorBase.errorType}
	 */
	readonly errorType = DriverErrorTypes.genericNetworkError;
```

(Same with the others)

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts
Comment (Josmithr) [inline packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:111]: We should be able to remove this now, right? Or is the flag itself deprecated?

**Review by Josmithr — CHANGES_REQUESTED** [review]
These changes are looking great! Thank you!!

My only pieces of feedback:
1. The new "Types" enums for `routerlicious-driver` and `odsp-driver` now _extend_ `DriverErrorTypes`. So for the most part, we can replace direct usages of `DriverErrorTypes.Foo` with the package-local error types. I think this would be a nice improvement for legibility of the code (so devs don't have to try to figure out the difference between the two types when reading).
2. Where possible, we should add `{@inheritDoc Foo}` comments to the `errorType` properties of error implementations so the generated docs pick up the documentation on the interface definition.

The only other issue here is timing. I think we need to wait for another internal major release to go out before we can merge this, that way Loop (and our other partners) have ample time to see and address the deprecation notices we left before having to deal with the breaking changes.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jikim-msft) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts]: Changed`DriverErrorTypes` to `OdspErrorTypes`.

File: packages/common/driver-definitions/src/driverError.ts
Comment (Josmithr) [inline packages/common/driver-definitions/src/driverError.ts:112]: Good call!

File: packages/drivers/odsp-driver-definitions/api-report/odsp-driver-definitions.api.md
Comment (Josmithr) [inline packages/drivers/odsp-driver-definitions/api-report/odsp-driver-definitions.api.md:223]: For visibility, this is a known API-Extractor issue. So long as CI generates the same content, this isn't an issue.

File: packages/drivers/odsp-driver-definitions/src/factory.ts
Comment (Josmithr) [inline packages/drivers/odsp-driver-definitions/src/factory.ts:10]: Nit:
```suggestion

	/**
```

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (alexvy86) [inline packages/drivers/odsp-driver/src/test/odspError.spec.ts:202]: 👀

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (alexvy86) [inline packages/drivers/odsp-driver/src/test/odspError.spec.ts:35]:
```suggestion
	 * Checks if the input is an {@link IGenericNetworkError}.
```

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (alexvy86) [inline packages/drivers/odsp-driver/src/test/odspError.spec.ts:45]:
```suggestion
	 * Checks if the input is an {@link IThrottlingWarning}.
```

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (alexvy86) [inline packages/drivers/odsp-driver/src/test/odspError.spec.ts:39]: Could this just compare `input.errorType` to `DriverErrorTypes.genericNetworkError`, without `typeof`? Same for the other helper function.

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (alexvy86) [inline packages/drivers/odsp-driver/src/test/odspError.spec.ts:106]: I think this should be enough, since the function already checks the `errorType`. If that's true (I didn't miss something), let's update the pattern in the other tests that we updated.

Also, can you add a note in the PR description regarding why we had to do this? I know we discussed it... and don't entirely recall the details, thus the ask 😅 .

```suggestion
		if (!isIGenericNetworkError(networkError)) {
```

**Review by alexvy86 — COMMENTED** [review]
@Josmithr this whole change should be only about types and have no runtime impact, right? I was trying to figure out if we needed to be careful if things with multiple versions might interact.

**Comment by Josmithr** [comment]
> @Josmithr this whole change should be only about types and have no runtime impact, right? I was trying to figure out if we needed to be careful if things with multiple versions might interact.

Correct - these should be type-only changes. The values of the underlying flags have not been modified.

File: docs/package.json
Comment (Josmithr) [inline docs/package.json:45]:
```suggestion
		"@microsoft/api-extractor-model": "7.27.6", 
```

File: packages/drivers/routerlicious-driver/src/test/errorUtils.spec.ts
Comment (jikim-msft) [reply packages/drivers/routerlicious-driver/src/test/errorUtils.spec.ts]: Removed 😃

File: docs/package.json
Comment (Josmithr) [inline docs/package.json:27]: This isn't related to your changes and should probably be reverted.

File: docs/package.json
Comment (Josmithr) [inline docs/package.json:42]: I would revert this so you can pick up the correct changes from `next` when you next merge it into this branch.

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (Josmithr) [inline packages/drivers/odsp-driver/src/test/odspError.spec.ts:46]:
```suggestion
	 */
```

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (Josmithr) [inline packages/drivers/odsp-driver/src/test/odspError.spec.ts:37]: I would recommend using `unknown` instead of `any` in these new functions. You'll also have to do something like the following in the code:
```typescript
(input as Partial<IGenericNetworkError>).errorType === ...
```

Comment (Josmithr) [reply packages/drivers/odsp-driver/src/test/odspError.spec.ts:37]: As a general rule, we should almost never use `any`.

File: packages/drivers/routerlicious-driver/src/test/errorUtils.spec.ts
Comment (Josmithr) [inline packages/drivers/routerlicious-driver/src/test/errorUtils.spec.ts:244]: This will result in no asserts being called if the error _isn't_ `IThrottlingWarning`. Rather than an `if`, we need to make this another `assert` call. I.e.
```typescript
assert(isIThrottlingWarning(error));
```

I _believe_ that should do the type-narrowing we need. If not, you can cast to the interface afterwards.

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (Josmithr) [inline packages/drivers/odsp-driver/src/test/odspError.spec.ts:161]: This shouldn't be necessary. The `isIGenericNetworkError` check should be sufficient.
```suggestion
```

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (Josmithr) [inline packages/drivers/odsp-driver/src/test/odspError.spec.ts:193]:
```suggestion
```

File: packages/drivers/routerlicious-driver/src/test/errorUtils.spec.ts
Comment (Josmithr) [inline packages/drivers/routerlicious-driver/src/test/errorUtils.spec.ts:20]: Let's use `unknown` here instead of `any` too.

File: packages/loader/container-loader/README.md
Comment (Josmithr) [inline packages/loader/container-loader/README.md:175]: I think these other links need to have their text changed too, since we renamed them as well.

File: packages/utils/odsp-doclib-utils/src/test/odspErrorUtils.spec.ts
Comment (Josmithr) [inline packages/utils/odsp-doclib-utils/src/test/odspErrorUtils.spec.ts:39]: This appears to be a new addition. Is this needed?

**Review by Josmithr — CHANGES_REQUESTED** [review]
I left a few more comments that I think need to be addressed before we can move forward. Looking close though!

File: docs/package.json
Comment (alexvy86) [inline docs/package.json:45]: Remind me why this change?

File: packages/drivers/odsp-driver-definitions/package.json
Comment (alexvy86) [inline packages/drivers/odsp-driver-definitions/package.json:59]: This bit seems unnecessary, the entry beneath it should be enough, I think.

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (alexvy86) [reply packages/drivers/odsp-driver/src/test/odspError.spec.ts:190]: Still see this console.log

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (alexvy86) [inline packages/drivers/odsp-driver/src/test/odspError.spec.ts:41]: If the function does happen to receive undefined, the current code will throw because it will try to access `errorType` on `undefined` before checking if `input` is `undefined`. We also don't need to cast to do that check.
```suggestion
			input !== undefined &&
			(input as Partial<IGenericNetworkError>).errorType ===
				DriverErrorTypes.genericNetworkError
```

File: packages/drivers/odsp-driver/src/test/odspError.spec.ts
Comment (alexvy86) [inline packages/drivers/odsp-driver/src/test/odspError.spec.ts:51]: Same as above.
```suggestion
			input !== undefined &&
			(input as Partial<IThrottlingWarning>).errorType === FluidErrorTypes.throttlingError
```

File: packages/drivers/routerlicious-driver/src/test/errorUtils.spec.ts
Comment (alexvy86) [inline packages/drivers/routerlicious-driver/src/test/errorUtils.spec.ts:24]: For consistency with the others :)
```suggestion
		return (
			input !== undefined &&
			(input as Partial<IThrottlingWarning>).errorType === FluidErrorTypes.throttlingError &&
			(input as Partial<IThrottlingWarning>).retryAfterSeconds !== undefined
		);
```

File: packages/utils/odsp-doclib-utils/src/test/odspErrorUtils.spec.ts
Comment (alexvy86) [inline packages/utils/odsp-doclib-utils/src/test/odspErrorUtils.spec.ts:54]: Since this is the only place where we used the function, I think I'm leaning towards removing it, removing the new assert, and where we need to check the `statusCode` do `assert.equal((networkError as Partial<IGenericNetworkError>).statusCode, 400, "status code should be preserved");`

File: docs/package.json
Comment (jikim-msft) [reply docs/package.json:45]: [Pipeline Failure: Build site artifact](https://github.com/microsoft/FluidFramework/actions/runs/6642240635/job/18046543552?pr=17651)

--

There was a bug in the `api-extractor` which created the pipeline failure above. @Josmithr suggested installing `"@microsoft/api-extractor/model": "7.27.6"` instead to bypass the error. 

It's been a while since I made changes to this PR, so I can check if the build passes after reverting this change.

File: docs/package.json
Comment (jikim-msft) [inline docs/package.json:74]: @Josmithr 

I think this is the reason that the PR validator failed.

Comment (jikim-msft) [reply docs/package.json:74]: NVM, failing with the same reason again 😓 .

**Review by Josmithr — APPROVED** [review]
Changes look good! But be sure to add a changeset that describes the breaking changes we are making here.

File: .changeset/mean-mugs-suffer.md
Comment (Josmithr) [inline .changeset/mean-mugs-suffer.md:20]: Nit: these changesets are intended for external consumers, so I don't think we need to call out test changes here.
```suggestion
The changes involve deprecation of error related enums ContainerErrorType, DriverErrorType, OdspErrorType and RouterliciousErrorType and replacement to new object-based enumerations of ContainerErrorTypes, DriverErrorTypes, OdspErrorTypes and RouterliciousErrorTypes.
```

Comment (alexvy86) [reply .changeset/mean-mugs-suffer.md:20]: I think the changeset should just say "Many deprecated classes are now removed. Refer to release notes of version <some_version> for the replacements." Or actually, I think this is one of the changes that we failed to announce before, so maybe the guidance on what to do to move away from the deprecated objects might have to go here.

But mainly my point is that this change doesn't deprecate anything; it removes things that were deprecated in the past.

And I agree with Josh's point about the tests, not necessary for release notes.

File: packages/drivers/routerlicious-driver/src/test/errorUtils.spec.ts
Comment (alexvy86) [inline packages/drivers/routerlicious-driver/src/test/errorUtils.spec.ts:247]: We can remove this line; `isIThrottlingWarning` already checks the value of `errorType`.

**Comment by anthony-murphy** [comment]
@jikim-msft i'm pretty sure next is permanently closed. you need to make this change in main. @tylerbutler  to confirm, and see if we can block checkins to next.

**Comment by jikim-msft** [comment]
@anthony-murphy 

I created a PR which makes the change from `next` to `main`. Just to confirm, should I revert the commit on the `next` branch as it is permanently closed?

**Comment by anthony-murphy** [comment]
> @anthony-murphy
> 
> I created a PR which makes the change from `next` to `main`. Just to confirm, should I revert the commit on the `next` branch as it is permanently closed?

it probably doesn't matter

**Pull request merged**
